src/hostapi/pulseaudio/pa_linux_pulseaudio_block.c

```diff
@@ -0,0 +1,186 @@
+ * license above.
+ */
+
+/** @file
+ @ingroup common_src
+
+ @brief PulseAudio implementation of support for a host API.
+
+ This host API implements PulseAudio support for portaudio
+ it has callbackmode and normal write mode support
+*/
+
+#include "pa_linux_pulseaudio_block_internal.h"
+#include <unistd.h>
+
+/*
+    As separate stream interfaces are used for blocking and callback
+    streams, the following functions can be guaranteed to only be called
+    for blocking streams.
+*/
+
+PaError PaPulseAudio_ReadStreamBlock( PaStream * s,
+                                      void *buffer,
+                                      unsigned long frames )
+{
+    PaPulseAudio_Stream *l_ptrStream = (PaPulseAudio_Stream *) s;
+    PaPulseAudio_HostApiRepresentation *l_ptrPulseAudioHostApi =
+        l_ptrStream->hostapi;
+    PaError l_iRet = 0;
+    size_t l_lReadable = 0;
+    uint8_t *l_ptrData = (uint8_t *) buffer;
+    long l_lLength = (frames * l_ptrStream->inputFrameSize);
+
+    pa_threaded_mainloop_lock( l_ptrStream->mainloop );
+    while( l_lLength > 0)
+    {
+        long l_read = PaUtil_ReadRingBuffer( &l_ptrStream->inputRing, l_ptrData,
+                                             l_lLength );
+        l_ptrData += l_read;
+        l_lLength -= l_read;
+        if( l_lLength > 0 )
+            pa_threaded_mainloop_wait( l_ptrStream->mainloop );
+    }
+    pa_threaded_mainloop_unlock( l_ptrStream->mainloop );
+    return paNoError;
+}
+
+
+PaError PaPulseAudio_WriteStreamBlock( PaStream * s,
+                                       const void *buffer,
+                                       unsigned long frames )
+{
+    PaPulseAudio_Stream *l_ptrStream = (PaPulseAudio_Stream *) s;
+    PaPulseAudio_HostApiRepresentation *l_ptrPulseAudioHostApi =
+        l_ptrStream->hostapi;
+    int l_iRet = 0;
+    size_t l_lWritable = 0;
+    uint8_t *l_ptrData = (uint8_t *) buffer;
+    long l_lLength = (frames * l_ptrStream->outputFrameSize);
+    pa_operation *l_ptrOperation = NULL;
+
+    PaUtil_BeginCpuLoadMeasurement( &l_ptrStream->cpuLoadMeasurer );
+
+    while( l_lLength > 0)
+    {
+        pa_threaded_mainloop_lock( l_ptrStream->mainloop );
+        l_lWritable = pa_stream_writable_size( l_ptrStream->outStream );
+        pa_threaded_mainloop_unlock( l_ptrStream->mainloop );
+
+        if( l_lWritable > 0 )
+        {
+           if( l_lLength < l_lWritable )
+           {
+                l_lWritable = l_lLength;
+           }
+           pa_threaded_mainloop_lock( l_ptrStream->mainloop );
+           l_iRet = pa_stream_write( l_ptrStream->outStream,
+                                     l_ptrData,
+                                     l_lWritable,
+                                     NULL,
+                                     0,
+                                     PA_SEEK_RELATIVE );
+
+            l_ptrOperation = pa_stream_update_timing_info( l_ptrStream->outStream,
```

Comment:
This can return 0 and let's the following code crash due to assertions in the pulse code.

Reply:
If it returns NULL it will be handled now